File: qabelbox/src/main/java/de/qabel/qabelbox/communication/connection/ConnectivityManager.java
package de.qabel.qabelbox.communication.connection;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;

public class ConnectivityManager {

    private Context context;

    private android.net.ConnectivityManager connectivityManager;
    private BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (listener != null) {
                if (isConnected()) {
                    listener.handleConnectionEtablished();
                } else {
                    listener.handleConnectionLost();
                }
            }
        }
    };

    protected ConnectivityListener listener;

    public ConnectivityManager(Context context) {
        this.context = context;
        this.connectivityManager = (android.net.ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);

        context.registerReceiver(broadcastReceiver, new IntentFilter(android.net.ConnectivityManager.CONNECTIVITY_ACTION));
    }

    public void onDestroy() {
        if(this.listener != null){
            this.listener.onDestroy();
        }
        context.unregisterReceiver(broadcastReceiver);
    }

    public boolean isConnected() {
        return connectivityManager.getActiveNetworkInfo() != null && connectivityManager.getActiveNetworkInfo().isConnected();
    }

    public boolean isWifi() {
        return connectivityManager.getActiveNetworkInfo().getType() == android.net.ConnectivityManager.TYPE_WIFI;
    }

    public void setListener(ConnectivityListener listener) {
        this.listener = listener;
    }

    public interface ConnectivityListener {

        void handleConnectionLost();

        void handleConnectionEtablished();

        void onDestroy();
    }
}

File: qabelbox/src/main/java/de/qabel/qabelbox/communication/connection/ConnectivityManager.kt
package de.qabel.qabelbox.communication.connection

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import rx.Observable
import rx.subjects.BehaviorSubject
import java.util.concurrent.TimeUnit

open class ConnectivityManager(private val context: Context) {

    private val connectivityManager: android.net.ConnectivityManager?

    private val connectedSubject: BehaviorSubject<Boolean> = BehaviorSubject.create<Boolean>()

    private val connected: Observable<Boolean> = connectedSubject.debounce(250, TimeUnit.MILLISECONDS)

    private val broadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            connectedSubject.onNext(isConnected)
        }
    }

    var listener: ConnectivityListener? = null

    init {
        this.connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE)
                as android.net.ConnectivityManager
        context.registerReceiver(broadcastReceiver,
                IntentFilter(android.net.ConnectivityManager.CONNECTIVITY_ACTION))
        connected.subscribe { connected ->
            if (connected) {
                listener?.handleConnectionEstablished()
            } else {
                listener?.handleConnectionLost()
            }
        }
    }

    fun onDestroy() {
        listener?.onDestroy()
        context.unregisterReceiver(broadcastReceiver)
    }

    open val isConnected: Boolean
        get() = connectivityManager?.activeNetworkInfo?.isConnected ?: false

    val isWifi: Boolean
        get() = connectivityManager?.activeNetworkInfo?.type == android.net.ConnectivityManager.TYPE_WIFI


    interface ConnectivityListener {

        fun handleConnectionLost()

        fun handleConnectionEstablished()

        fun onDestroy()
    }
}
